src/abstractdataframe/reshape.jl

```diff
@@ -185,39 +193,71 @@ long = stack(wide)
 wide0 = unstack(long)
 wide1 = unstack(long, :variable, :value)
 wide2 = unstack(long, :id, :variable, :value)
+wide3 = unstack(long, [:id, :a], :variable, :value)
 ```
 Note that there are some differences between the widened results above.
-
 """
-function unstack(df::AbstractDataFrame, rowkey::Int, colkey::Int, value::Int)
-    # `rowkey` integer indicating which column to place along rows
-    # `colkey` integer indicating which column to place along column headers
-    # `value` integer indicating which column has values
-    refkeycol = CategoricalArray{Union{eltype(df[rowkey]), Missing}}(df[rowkey])
-    # make sure we report only levels actually present in rowkey column
-    # this is consistent with how the other unstack method based on groupby works
+function unstack(df::AbstractDataFrame, rowkeys::Int, colkey::Int, value::Int)
+    refkeycol = copy(CategoricalArray{Union{eltype(df[rowkeys]), Missing}}(df[rowkeys]))
     droplevels!(refkeycol)
-    valuecol = df[value]
     keycol = CategoricalArray{Union{eltype(df[colkey]), Missing}}(df[colkey])
+    valuecol = df[value]
     Nrow = length(refkeycol.pool)
     Ncol = length(keycol.pool)
-    payload = DataFrame(Any[similar_missing(valuecol, Nrow) for i in 1:Ncol], map(Symbol, levels(keycol)))
-    nowarning = true
+    df2 = DataFrame(Any[similar_missing(valuecol, Nrow) for i in 1:Ncol], map(Symbol, levels(keycol)))
+    _unstack(df, rowkeys, colkey, value, keycol, valuecol, df2, refkeycol)
+end
+
+function _unstack(df::AbstractDataFrame, rowkeys::Int,
+                  colkey::Int, value::Int, keycol, valuecol, df2, refkeycol)
+    Nrow = length(refkeycol.pool)
+    Ncol = length(keycol.pool)
+    hadmissing = false # have we encounered missing in refkeycol
+    mask_filled = falses(Ncol, Nrow) # has a given [col,row] entry been filled?
+    nowarning = true # do we print duplicate entries warning
+    nowarning2 = true # do we print missing in keycol
+    keycol_pool = Vector{Int}(CategoricalArrays.order(keycol.pool))
+    refkeycol_pool = Vector{Int}(CategoricalArrays.order(refkeycol.pool))
     for k in 1:nrow(df)
-        j = Int(CategoricalArrays.order(keycol.pool)[keycol.refs[k]])
-        i = Int(CategoricalArrays.order(refkeycol.pool)[refkeycol.refs[k]])
-        if i > 0 && j > 0
-            if nowarning && !ismissing(payload[j][i])
-                warn("Duplicate entries in unstack.")
-                nowarning = false
+        keycol_refs = keycol.refs[k]
+        if keycol_refs == 0 # we have found missing in colkey
+            if nowarning2
+              warn("Missing value in colkey variable at row $k. Skipping.")
```

Comment:
Indentation issue. Would also be nice to print the name of the variable while you're at it.

Reply:
fixed